File: python_implementation/bruteforce.py
import itertools
from typing import List, Tuple
from typehints import Action


def bruteforce(actions: List[Action], max_weight: int) -> Tuple[List[str], float]:

    all_combinations = []
    for r in range(1, len(actions) + 1):
        combinations_object = itertools.combinations(actions, r)
        all_combinations += combinations_object

    possible_solutions = []
    for combination in all_combinations:
        weight = sum(item[1] for item in combination)
        if weight <= max_weight:
            value = sum(item[1] * item[2] for item in combination)
            possible_solutions.append((combination, value))

    best_solution = max(possible_solutions, key=lambda x: x[1])
    best_solution_actions = [item[0] for item in best_solution[0]]
    best_solution_value = best_solution[1]

    return best_solution_actions, best_solution_value

File: rust_implementation/src/bruteforce.rs
use itertools::Itertools;
use crate::models::Action;
use std::cmp::Ordering;

pub fn bruteforce(actions: &[Action], max_weight: i32) -> (Vec<&Action>, f64) {
    let mut all_combinations = vec![];
    for i in 1..=actions.len() {
        let combinations: Vec<Vec<&Action>> = actions.iter().combinations(i).collect();
        all_combinations.extend(combinations);
    }

    let mut possible_solutions = vec![];
    for combination in &all_combinations {
        let weight: i32 = combination.iter().map(|action| action.cost).sum();
        if weight <= max_weight {
            let value: f64 = combination.iter().map(|action| action.cost as f64 * action.value).sum();
            possible_solutions.push((combination, value));
        }
    }

    let (best_solution, best_value) = possible_solutions
        .into_iter()
        .max_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Equal))
        .unwrap();

    let best_solution_actions: Vec<&Action> = best_solution.to_vec();

    (best_solution_actions, best_value)
}